clippy_lints/src/const_static_lifetime.rs

```diff
@@ -0,0 +1,87 @@
+use syntax::ast::{Item, ItemKind, TyKind, Ty};
+use rustc::lint::{LintPass, EarlyLintPass, LintArray, EarlyContext};
+use utils::{span_lint_and_then, in_macro, snippet};
+
+/// **What it does:** Checks for constants with an explicit `'static` lifetime.
+///
+/// **Why is this bad?** Adding `'static` to every reference can create very
+/// complicated types.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// const FOO: &'static [(&'static str, &'static str, fn(&Bar) -> bool)] =
+/// &[...]
+/// ```
+/// This code can be rewritten as
+/// ```rust
+///  const FOO: &[(&str, &str, fn(&Bar) -> bool)] = &[...]
+/// ```
+
+declare_lint! {
+    pub CONST_STATIC_LIFETIME, 
+    Warn,
+    "Using explicit `'static` lifetime for constants when elision rules would allow omitting them."
+}
+
+pub struct StaticConst;
+
+impl LintPass for StaticConst {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(CONST_STATIC_LIFETIME)
+    }
+}
+
+impl StaticConst {
+    // Recursively visit types
+    fn visit_type(&mut self, ty: &Ty, cx: &EarlyContext) {
+        match ty.node {
+            // Be carefull of nested structures (arrays and tuples)
+            TyKind::Array(ref ty, _) => {
+                self.visit_type(&*ty, cx);
+            },
+            TyKind::Tup(ref tup) => {
+                for tup_ty in tup {
+                    self.visit_type(&*tup_ty, cx);
+                }
+            },
+            // This is what we are looking for !
+            TyKind::Rptr(ref optional_lifetime, ref borrow_type) => {
+                // Match the 'static lifetime
+                if let Some(lifetime) = *optional_lifetime {
+                    if let TyKind::Path(_, _) = borrow_type.ty.node {
+                        // Verify that the path is a str
+                        if lifetime.ident.name == "'static" {
+                            let mut sug: String = String::new();
+                            // This should be ok, since we know that the current type is a borrow.
+                            sug.push('&');
+                            sug += &snippet(cx, borrow_type.ty.span, "<expr>").into_owned();
```

Comment:
The suggestion should be empty. It is the code that is written over `'static`, not over `&'static T`.

Reply:
Oh, thanks I couldn't figured it out.
Is this for the application that automatically applies suggestions ?

Reply:
Yea. In the future rls will probably make use of it, too. So you'll be able to fix these things with a single click on the offending error message.

Reply:
I am currently applying your requests, but `cargo test` fails because :
```
error: Unnecessary `!=` operation
   --> clippy_lints/src/loops.rs:754:21
    |
754 | /                     if offset_left.var_name != offset_right.var_name {
755 | |                         Some((offset_left, offset_right))
756 | |                     } else {
757 | |                         None
758 | |                     }
    | |_____________________^
    |
    = note: `-D if-not-else` implied by `-D clippy-pedantic`
    = help: change to `==` and swap the blocks of the if/else
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.166/index.html#if_not_else

error: aborting due to previous error
```
Is this related to the latest nigthly or some error in my changes ?

Reply:
curious. I wonder why travis didn't catch that on the other PRs... I'll push a quick fix that you can rebase over

Reply:
You can rebase over the master branch, I fixed it